document/JS.bs

```diff
@@ -0,0 +1,935 @@
+    1. Assert: |ret| is empty.
+    1. Let |exportsObject| be  ! ObjectCreate(null).
+    1. For each pair (|name|, |externtype|) in [=module_exports=](|module|),
+      1. Let |externval| be [=get_export=](|instance|, |name|).
+      1. If |externtype| is of the form 𝖿𝗎𝗇𝖼 |functype|,
+         1. If |funcs| contains an entry |func| where |func|.\[[FunctionAddress]] is |externval|, let |value| be |func|
+         1. Otherwise,
+            1. Let |func| be the result of [=create an Exported Function=] from |externval|.
+            1. Append |func| to |funcs|.
+            1. Let |value| be |func|.
+      1. If |externtype| is of the form 𝗀𝗅𝗈𝖻𝖺𝗅 |globaltype|,
+         1. If |globaltype|.valtype is an [=i64=], throw a {{LinkError}} exception.
+         1. Assert: |globaltype|.mut is const.
+         1. Let |value| be [=ToJSValue=]([=read_global=](|store|, |externval|)).
+      1. If |externtype| is of the form 𝗆𝖾𝗆  |memtype|,
+         1. If there is an element |memory| in |memories| such that |memory|.\[[Memory]] is |externval|, then let |value| be |memory|
+         1. Otherwise:
+            1. Let |memory| be [=create a memory object|a new Memory object=] from |externval|.
+            1. Append |memory| to |memories|.
+            1. Let |value| be |memory|.
+      1. Otherwise, |externtype| is of the form 𝗍𝖺𝖻𝗅𝖾 |tabletype|,
+         1. Let |value| be the unique |table| in |tables| such that |table|.\[[Table]] is |externval|.
+      1. Perform ! CreateDataPropertyOrThrow(|exports|, |name|, |value|).
+
+      Note: the validity and uniqueness checks performed during [=WebAssembly module validation=] ensure that each property name is valid and no properties are defined twice.
+
+    33. Perform ! [=SetIntegrityLevel=](|exportsObject|, `"frozen"`).
+    1. Let |instanceObject| be a new {{Instance|| object setting whose internal \[[Instance]] slot is set to |instance| and the \[[Exports]] slot to |exportsObject|.
+    1. Return |instanceObject|.
+</div>
+
+<div algorithm>
+  To <dfn>instantiate a promise of a module</dfn> |promiseOfModule| with imports |importObject|, perform the following steps:
+
+    1. Let |promise| be [=a new promise=]
+    1. [=Upon fulfillment=] of |promiseOfModule| with value |module|:
+        1. [=instantiate a WebAssembly module|Instantiate the WebAssembly module=] |module| importing |importObject|, and let |instance| be the result.  If this throws an exception, catch it, [=reject=] |promise| with the exception, and abort these substeps.
+        1. Let |result| be a {{WebAssemblyInstantiatedSource}} dictionary with {{WebAssemblyInstantiatedSource/module}} set to |module| and {{WebAssemblyInstantiatedSource/instance}} set to |instance|.
+        1. [=Resolve=] |promise| with |result|.
+        1. Other unspecified asynchronous tasks may run at this point.
```

Comment:
What is the purpose of this step?

Reply:
This was added in https://github.com/WebAssembly/design/commit/3fb9bce77588f45912e791f60daede1183112772 . From the commit message,

> Additionally, an implementor may wish to do more asynchronous work after they have looked at a module's imports. This change also explicitly allows implementors to do more asynchronous work.

Is this implicit, and the spec line not needed? If so, maybe the right thing to do is make it a note.

Reply:
Oh, that is weird. I think it should be deleted entirely. An implementation may always do any unobservable asynchronous work it wants. Otherwise we'd put a similar note after every queued task, or indeed every algorithm step. Maybe I am missing something.

Reply:
@kmiller68 What do you think? You wrote the previous text.